numenc/encoder_decoder.c

```diff
@@ -0,0 +1,947 @@
+    }
+
+    if (endianness == -1) {
+        endianness = endianness_impl();
+    }
+
+    char * buffer = (char * ) PyMem_RawMalloc(EIGHT_BYTES);
+    if (buffer == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    if (endianness == 0) {
+        for (int i = 0; i < EIGHT_BYTES; i++) {
+            buffer[i] = (input >> (i * 8)) & 0xff;
+        }
+    } else {
+        for (int i = 0; i < EIGHT_BYTES; i++) {
+            buffer[EIGHT_BYTES - 1 - i] = (input >> (i * 8)) & 0xff;
+        }
+    }
+
+    output = PyBytes_FromStringAndSize(buffer, EIGHT_BYTES);
+    PyMem_RawFree(buffer);
+    return output;
+}
+
+static PyObject * to_uint64(PyObject * self, PyObject * args) {
+    const char * input;
+    PyObject * output;
+    int count;
+
+    if (!PyArg_ParseTuple(args, "y#", & input, & count)) {
+        return PyErr_Format(PyExc_TypeError, "Wrong input: expected bytes, but got %s", Py_TYPE(args) -> tp_name);
+    }
+
+    if (endianness == -1) {
+        endianness = endianness_impl();
+    }
+
+    if (count != EIGHT_BYTES) {
+        return PyErr_Format(PyExc_ValueError, "Illegal input: expected bytes of length %d, got %d.",
+                            EIGHT_BYTES, count);
+    }
+
+    unsigned char * buffer = (unsigned char * ) PyMem_RawMalloc(EIGHT_BYTES);
+    if (buffer == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    for (int i = 0; i < EIGHT_BYTES; i++) {
+        buffer[i] = (unsigned char) input[i];
+    }
+
+    uint64_t decoded = 0;
+    if (endianness == 0) {
+        for (int i = 0; i < EIGHT_BYTES; i++) {
+            decoded |= ((uint64_t) buffer[i]) << (i * 8);
+        }
+    } else {
+        for (int i = 0; i < EIGHT_BYTES; i++) {
+            decoded |= ((uint64_t) buffer[EIGHT_BYTES - 1 - i]) << (i * 8);
+        }
+    }
+    output = PyLong_FromUnsignedLongLong(decoded);
+    PyMem_RawFree(buffer);
+    return output;
+}
+
+static PyObject * from_float32(PyObject * self, PyObject * args) {
+    float input;
+    PyObject * output;
+
+    if (!PyArg_ParseTuple(args, "f", & input)) {
+        return PyErr_Format(PyExc_TypeError, "Wrong input: expected 32-bit float.");
+    }
+
+    if (endianness == -1) {
+        endianness = endianness_impl();
+    }
+
+    union float_bytes f_bytes;
+    f_bytes.val = input;
+
+    char * buffer = (char * ) PyMem_RawMalloc(FOUR_BYTES);
+    if (buffer == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    // positive number: set sign bit to 1
+    if (input >= 0) {
+        if (endianness == 0) {
+            for (int i = 0; i < FOUR_BYTES; i++) {
+                buffer[i] = f_bytes.bytes[i];
+            }
+        } else {
+            for (int i = 0; i < FOUR_BYTES; i++) {
+                buffer[FOUR_BYTES - 1 - i] = f_bytes.bytes[i];
+            }
+        }
+        buffer[0] |= 0x80;
+        // negative number: flip all bits
```

Comment:
This comment should go below to else, shouldn't it? Otherwise, I'd think that you forgot to add implementation here.

Reply:
yes, it goes in the `else`